src/train_cmd.cpp

```diff
@@ -959,6 +1024,83 @@ static CommandCost CheckNewTrain(Train *original_dst, Train *dst, Train *origina
 	return_cmd_error(STR_ERROR_TOO_MANY_VEHICLES_IN_GAME);
 }
 
+/**
+ * Check if a train can be placed in a depot tile.
+ * @param train The train.
+ * @param tile The tile to check whether it is possible to place the train.
+ * @return whether it found a depot tile to place the train.
+ */
+bool CheckPlacement(const Train *train, TileIndex tile)
+{
+	assert(train != nullptr);
+	assert(IsRailDepotTile(tile));
+
+	RailType rt = GetRailType(tile);
+	for (const Train *t = train; t != nullptr; t = t->Next()) {
+		RailType rail_type = Engine::Get(t->engine_type)->u.rail.railtype;
+		if (!IsCompatibleRail(rail_type, rt)) return false;
+	}
+
+	return true;
+}
+
+/**
+ * Before placing a train in the rails of a depot, a valid depot tile must
+ * be found. This function finds a tile for placing the train.
+ * @param t The train to place in a rail depot tile.
+ * @return a compatible tile, if any, preferabily the one the first vehicle is or INVALID_TILE if none found.
+ */
+TileIndex LookForTileInDepot(const Train *train)
+{
+	assert(train != nullptr);
+	assert(IsRailDepotTile(train->tile));
+	TileIndex best_tile = INVALID_TILE;
+
+	/* First candidate is the original position of the train. */
+	if (CheckPlacement(train, train->tile)) {
+		if (HasPowerOnRail(train->railtype, GetRailType(train->tile))) return train->tile;
+		best_tile = train->tile;
+	}
+
+	/* Check all depot tiles. */
+	Depot *depot = Depot::GetByTile(train->tile);
+	for (std::vector<TileIndex>::iterator it = depot->depot_tiles.begin(); it != depot->depot_tiles.end(); ++it) {
+		if (CheckPlacement(train, *it)) {
+			if (HasPowerOnRail(train->railtype, GetRailType(*it))) return *it;
+			if (best_tile == INVALID_TILE) best_tile = *it;
+		}
+	}
+
+	return best_tile;
+}
+
+/**
+ * Find an appropriate depot tile for a train and place
+ * all the vehicle chain in the same depot tile.
+ * @param train The train to place.
+ */
+void PlaceOnRailDepot(Train *train)
+{
+	assert(train->First() == train);
+	TileIndex depot_tile = LookForTileInDepot(train);
+
+	if (depot_tile == INVALID_TILE) NOT_REACHED();
```

Comment:
Is there a reason this is a `NOT_REACHED()` instead of an `assert()`?

```suggestion
	assert(depot_tile == INVALID_TILE);
```

Reply:
The suggested change inverts the meaning of the test.

Reply:
Oof, it's been a long day. Edited.